Add code needed for upcoming "Progressive Jump" option

soopercool101 opened the pull request:
Implements the ID previously reserved for Double Jump as "Progressive Jump". Similar to how Progressive Key works, this item will give Double Jump upon receiving one of these items, and Triple Jump upon receiving a second. This client is backwards-compatible with current jump behavior, by giving both Double and Triple Jump whenever the Triple Jump item is received (this is why it can be merged directly into the `archipelago` branch rather than needing to be in `ap_test`). Tested and confirmed working on both the current APWorld and the Beta currently found in the Move Rando thread, and works with all officially supported patches.

soopercool101 committed:
Initial changes needed for Double Jump

soopercool101 committed:
Fix menu spacing to accommodate new move

soopercool101 committed:
Skip reading DJ bit

Should fix issues with Double Jump always appearing with progressive jump off and on older APWorlds

N00byKing commented on src/sm64ap.cpp:
```diff
@@ -79,8 +79,12 @@ void SM64AP_RecvItem(int64_t idx, bool notify) {
+            sm64_have_abilities[idx - SM64AP_ABILITY_OFFSET] = true;
+            break;
+        case SM64AP_ID_ABILITY(1) ... SM64AP_ID_ABILITY(SM64AP_NUM_ABILITIES-1):
+            sm64_have_abilities[idx-SM64AP_ABILITY_OFFSET] = true;
```
Spacing

soopercool101 replied:
Fixed in 9746f720e2f38a8bdb6961b735559cd705ac381f

N00byKing commented on src/sm64ap.cpp:
```diff
@@ -267,7 +271,7 @@ void SM64AP_SetCourseMap(std::map<int,int> map) {
 
 void SM64AP_SetMoveRandoVec(int vec) {
-    for (int i = 0; i < SM64AP_NUM_ABILITIES; i++) {
+    for (int i = 1; i < SM64AP_NUM_ABILITIES; i++) { // Start at 1, DJ bit is unnecessary
```
Does this cause issues without this change? If not, keep as is

soopercool101 replied:
Without these changes, current APWorld will always give Double Jump, as the server's DJ bit is always set to 0. The design in the upcoming APWorld does not allow DJ to be set without TJ, so reading this bit is no longer necessary from the client end. @Magnemania confirmed this behavior in our Discord conversation [here](https://discord.com/channels/731205301247803413/1197074752687312917/1312591923311018077)

N00byKing commented on src/game/ingame_menu.c:
```diff
@@ -2714,49 +2714,54 @@ s16 render_pause_courses_and_castle(void) {
     if (SM64AP_CanTripleJump()) {
         u8 str_triple_jump[] = { TEXT_TRIPLE_JUMP };
-        print_generic_string(x, y, str_triple_jump);
+        print_generic_string(x, y - spacing, str_triple_jump);
```
Why is the spacing change necessary?

soopercool101 replied:
Adding Double Jump to the ability list results in the last move being cut off with the old spacing. Spacing of 16 was chosen as it results in roughly the same position for the final move (18\*9=162, 16\*10=160). Here is a screenshot of what the ability list looks like with this change applied:
![image](https://github.com/user-attachments/assets/6348462b-85cc-4da0-9287-cdc5b4e2709e)

soopercool101 replied:
Spacing of 17 was also considered, but I ultimately didn't like how close it moves the final ability to the bottom of the screen. Here is what that looks like, for reference:
![image](https://github.com/user-attachments/assets/77b0f986-ca53-4f12-b4c3-1d8cd1f3ba77)

soopercool101 replied:
I suppose I could also just have the original list and have the string that currently says Triple Jump change conditionally on the level of jump received. Let me know if that is a preferrable solution

N00byKing replied:
Yeah that sounds nice

soopercool101 replied:
Fixed in accf68bcea0e9c05fd3808c4b8d5a196b99d8cf1

N00byKing reviewed (commented):
Dont have time to test right now, so just looking over the code

soopercool101 committed:
Match formatting in SM64AP_RecvItem

soopercool101 committed:
Ability List: Show Double/Triple Jump in the same slot

Show Double Jump if that's all that's been received, Triple Jump if it's been received

N00byKing commented:
Thanks!
